src/main/java/com/n1analytics/paillier/EncryptedNumber.java

```diff
@@ -416,4 +416,28 @@ public boolean equals(EncryptedNumber o) {
             context.equals(o.context) &&
             ciphertext.equals(o.ciphertext));
   }
+
+  /**
+   * Return a byte array representing the underlying raw number.
+   *
+   * WARNING: This method may return unencrypted information! Under no
+   * circumstances should the return value be sent anywhere.
+   *
+   * WARNING: This method is not intended for general use and may be
+   * removed in a future release. Think twice before using it.
+   */
+  public byte[] toByteArrayUnsafe() {
+    return ciphertext.toByteArray();
+  }
+
+  /**
+   * Return true or false according to whether this EncryptedNumber is
+   * safe for transmission.
+   *
+   * WARNING: This method is not intended for general use and may be
+   * removed in a future release. Think twice before using it.
+   */
+  public boolean getIsSafe() {
```

Comment:
Right, so the serialisation you refer to occurs at step 3 (and in reverse at step 7) in the list I posted above. Indeed I am obliged to create at least (but I think _only_) two intermediate arrays to "serialise" the `BigInteger[]` for consumption by the GPU. Happily there isn't much in the way of 'massaging' to do, since the only oddities of the `BigInteger.toByteArray()` result are (i) it is big endian, and (ii) the "most-significant" byte in the resulting array is allowed to be zero (this is a weird normalisation, for handling signed BigIntegers I think).

I have measured the overhead of this array creation/massaging/copying at various times in the past and it has never been significant compared to the the other calculations, though this may have changed recently with the incorporation of Felix's trick, which has invalidated most of my previous performance numbers. With some extra work I believe I could reduce the serialisation I do to essentially a single copy from the JVM to the GPU memory, if this turns out to be a bottleneck.

The performance question is still more-or-less orthogonal to the problem that this PR was supposed to solve though, namely: at some point, somehow, I need to get access to the `ciphertext`s inside the `EncryptedNumber[]` so that I can send them to the GPU for processing.